storage/kafka/src/main/java/io/apicurio/registry/kafka/utils/ConsumerContainer.java

```diff
@@ -0,0 +1,408 @@
+
+                    if (activeTopics != null) {
+                        Collection<TopicPartition> idlePartitions = null;
+                        for (Map.Entry<TopicPartition, Long> e : activeTopics.entrySet()) {
+                            if (time - e.getValue() >= idlePingTimeout) {
+                                if (idlePartitions == null)
+                                    idlePartitions = new ArrayList<>(4);
+                                idlePartitions.add(e.getKey());
+                                e.setValue(time);
+                            }
+                        }
+                        if (idlePartitions != null) {
+                            idlePartitions.forEach(idlePingConsumer);
+                            activeTopics.keySet().retainAll(consumer.assignment());
+                        }
+                    }
+                }
+            }
+        } catch (Throwable e) {
+            log.warn("Exception caught in consumer polling thread", e);
+        } finally {
+            log.info("Consumer thread exiting");
+            running = false;
+        }
+    }
+
+    private void consume(ConsumerRecords<K, V> records, Consumer<K, V> consumer) {
+        if (recordsConsumer != null) {
+            consumeRetryable(records, recordsConsumer, consumer);
+        } else {
+            assert recordConsumer != null;
+            for (ConsumerRecord<K, V> record : records) {
+                consumeRetryable(record, recordConsumer, consumer);
+            }
+        }
+    }
+
+    /**
+     * Consume record by passing it to {@link #recordConsumer} and retrying that until successful (possibly ad infinitum)
+     *
+     * @param record the record to consume
+     */
+    private <T> void consumeRetryable(T record, java.util.function.Consumer<? super T> recordConsumer, Consumer<K, V> consumer) {
+        long delay = MIN_RETRY_DELAY;
+        boolean interrupted = false;
+        // retry loop
+        while (!stopping) try {
+            recordConsumer.accept(record);
+            // everything alright
+            break;
+        } catch (Exception e) {
+            log.warn("Exception caught while processing {} - retrying in {} ms", formatRecord(record), delay, e);
+            CompletableFuture<Consumer<K, V>> task;
+            try {
+                while (!stopping && (task = tasks.poll(delay, TimeUnit.MILLISECONDS)) != null) {
+                    task.complete(consumer);
+                }
+            } catch (InterruptedException ie) {
+                log.info("Interrupted - keeping retrying");
+                interrupted = true;
+            }
+            delay = Math.min(delay * 2, MAX_RETRY_DELAY);
+        } catch (Throwable e) {
+            // in case we get Error (OutOfMemoryError or similar) there's no point in continue-ing with retries, but we can't
+            // ignore the error either. So what we're left with is bailing-out and hoping that Kubernetes gives us a fresh start...
```

Comment:
Yeah, it can be non-Kuberenetes ...
What do you mean "better handled via -XX:+ExitOnOutOfMemoryError" ?

Reply:
It's a JVM option which can be used to exit the process when the JVM hits an OOM condition. It's probably more reliable than trying to code something special because you never know in which thread an OOME will be thrown.

Reply:
OK, yeah, that can be used additionally.
On any other error, there is also not much we can do, hence exit.